### DISCUSSION TITLE
Virtual Module Support

### DISCUSSION THREAD

**Discussion by TylorS** [volarjs/volar.js#35] [General]
Hey there, thanks for open-sourcing such a cool project. This is such a large barrier to entry for a lot of tools.

I've personally been building a vite-plugin that utilizes virtual modules to generate boilerplate-heavy code that can be automated. At the moment, I'm only generating TypeScript files, but I'm curious if I could/should use Volar to help the TypeScript language server/service understand the types of the generated files behind a virtual module import?

**Comment by johnsoncodehk** [comment]
You can, that's exactly what we do with .vue behind the scenes. The general idea is to generate a corresponding .xxx.d.ts virtual file for files that need to be recognized by TS.

Is your use case specific to vue project? Can you provide a more specific minimal case?

**Reply by TylorS** [reply]
Hey @johnsoncodehk Thanks for the reply. My use-case is outside of vue, and more built on [Effect](https://effect.website/). I actually have a POC example of what I mean over here - https://github.com/TylorS/typed-fp/blob/development/example/server.ts#LL6C1-L10C41. Basically the virtual module imports utilize prefixes like `api:` or `html:` and so on to determine what kind of content to generate. For `api:` for example, it traverses the supplied directory (or just a file) and uses TypeScript to statically analyze which exports are Handlers, and which are "environments" (Effect supports type-safe dependency injection), and generates the list of handlers with all of their environments provided topologically.

I currently have it built only with Vite's support for virtual modules and a [static .d.ts file](https://github.com/TylorS/typed-fp/blob/development/packages/framework/src/virtual-modules.d.ts) for the POC, but I'm hoping to improve the editing and type-checking experience by integrating with TypeScript as well.

**Reply by TylorS** [reply]
I've had some minor success with setting up my own language service and patching the language host [here](https://github.com/TylorS/typed-fp/blob/ts/feat/virtual-modules/packages/virtual/src/setupProject.ts#L46), and it's working for type-checking and for emits, but no matter what I seem to do, when I try to replicate it into a TS language service plugin it will always cause the server to crash without any error message so I started looking for anyone else doing this stuff and came across this project

**Reply by johnsoncodehk** [reply]
I haven't looked into it closely yet. If you just need to make a TS plugin, you can refer to typescript-plugin-css-modules, which has successfully achieved similar requirements.

https://github.com/mrmckeb/typescript-plugin-css-modules/blob/main/src/index.ts#L210

But Volar.js may not meet your use case, it is a framework for creating a complete LSP server and related tool chain, but not for TS plugin.

**Reply by johnsoncodehk** [reply]
We have a discord server, if you think you still need to research Volar.js, you may consider joining it and find me.

https://discord.gg/5bnSSSSBbK
